### PR TITLE
[USM] Support HTTP2 request cases

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

This PR supports `Literal Never Indexed` and `Literal Without Indexing` for HTTP/2 decoding.
While parsing the literal cases, there are two additional cases that we have not yet supported:

- Literal Header Field without Indexing, where the top four bits are 0000 (referenced in: [RFC7541 Section 6.2.2](https://httpwg.org/specs/rfc7541.html#rfc.section.6.2.2))

- Literal Header Field never Indexed, where the top four bits are 0001 (referenced in: [RFC7541 Section 6.2.3](https://httpwg.org/specs/rfc7541.html#rfc.section.6.2.3))

These cases were identified during the TLS support development and are pertinent to the following [commit](https://github.com/DataDog/datadog-agent/pull/21619/commits/64b801ede094d155bd8c6f627facb98b31ad8f9b).

### Motivation

Supports decoding all possible HTTP/2 cases to handle various types of traffic.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 1]
```diff
@@ -248,11 +255,18 @@ static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table
...
 
             // create the new dynamic value which will be added to the internal table.
             read_into_buffer_path(dynamic_value.buffer, skb, current_header->new_dynamic_value_offset);
+            // If the value is indexed - add it to the dynamic table.
+            if (current_header->type == kNewDynamicHeader) {
+                dynamic_value.string_len = current_header->new_dynamic_value_size;
+                dynamic_value.is_huffman_encoded = current_header->is_huffman_encoded;
+                bpf_map_update_elem(&http2_dynamic_table, dynamic_index, &dynamic_value, BPF_ANY);
+            }
             bpf_map_update_elem(&http2_dynamic_table, dynamic_index, &dynamic_value, BPF_ANY);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:270]: should be deleted as you added it in the `if (current_header->type == kNewDynamicHeader)`

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 2]
```diff
@@ -248,11 +255,18 @@ static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table
             current_stream->is_huffman_encoded = dynamic_value->is_huffman_encoded;
             bpf_memcpy(current_stream->request_path, dynamic_value->buffer, HTTP2_MAX_PATH_LEN);
         } else {
+            // We're in new dynamic header or new dynamic header not indexed states.
             dynamic_value.string_len = current_header->new_dynamic_value_size;
             dynamic_value.is_huffman_encoded = current_header->is_huffman_encoded;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:260]: should be deleted as you added it in the `if (current_header->type == kNewDynamicHeader)`

File: pkg/network/usm/monitor_test.go [hunk 3]
```diff
@@ -978,16 +978,31 @@ var (
...
 )
 
-// createRawRequestFrame creates a raw request frame with the given streamID.
-func createRawRequestFrame(streamID int) []byte {
-	binary.BigEndian.PutUint16(request[7:9], uint16(streamID))
-	return request
+const (
+	caseWithoutIndexing = 1
+	caseNeverIndexed    = 2
+)
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:984]:
```suggestion
type case int
const (
	caseWithoutIndexing case = iota
	caseNeverIndexed    case
)
```

File: pkg/network/usm/monitor_test.go [hunk 4]
```diff
@@ -978,16 +978,31 @@ var (
...
+
+// createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
+func createHTTP2RequestByCase(requestType int) []byte {
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:987]:
```suggestion
func createHTTP2RequestByCase(requestType case) []byte {
```

File: pkg/network/usm/monitor_test.go [hunk 5]
```diff
@@ -978,16 +978,31 @@ var (
...
+	switch requestType {
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		request[22] = 0x05
+		return request
+	// http2 request to path /aaa which never indexed (same as request but with different index changed from 0x45 to 0x15)
+	case caseNeverIndexed:
+		request[22] = 0x15
+		return request
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:996]: maybe it is better to write the binary mode

File: pkg/network/usm/monitor_test.go [hunk 6]
```diff
@@ -978,16 +978,31 @@ var (
...
+	caseNeverIndexed    = 2
+)
+
+// createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
+func createHTTP2RequestByCase(requestType int) []byte {
+	switch requestType {
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		request[22] = 0x05
+		return request
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:992]: redundant

File: pkg/network/usm/monitor_test.go (hunk 5 above)
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:996]: redundant

File: pkg/network/usm/monitor_test.go [hunk 7]
```diff
@@ -978,16 +978,31 @@ var (
...
+	// http2 request to path /aaa which never indexed (same as request but with different index changed from 0x45 to 0x15)
+	case caseNeverIndexed:
+		request[22] = 0x15
+		return request
+	default:
+		return request
+	}
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:999]:
```suggestion
	default:
	}
	return request
```

**Review by guyarb — CHANGES_REQUESTED** [review]
reviewed

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 1 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:270]: removed

File: pkg/network/usm/monitor_test.go [hunk 8]
```diff
@@ -964,39 +964,38 @@ var (
...
 // createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
-func createHTTP2RequestByCase(requestType int) []byte {
+func createHTTP2RequestByCase(requestType literalType) []byte {
+	// http2 request to path /aaa with index 5
+	request := []byte{
+		0x00, 0x00, 0x37, 0x01, 0x04, 0x00, 0x00, 0x00, 0x03, 0x41, 0x8a, 0x08, 0x9d, 0x5c, 0x0b, 0x81,
+		0x70, 0xdc, 0x78, 0x0f, 0x0b, 0x83, 0x45, 0x83, 0x60, 0x63, 0x1f, 0x86, 0x5f, 0x8b, 0x1d, 0x75,
+		0xd0, 0x62, 0x0d, 0x26, 0x3d, 0x4c, 0x74, 0x41, 0xea, 0x5c, 0x01, 0x34, 0x50, 0x83, 0x9b, 0xd9,
+		0xab, 0x7a, 0x8d, 0xc4, 0x75, 0xa7, 0x4a, 0x6b, 0x58, 0x94, 0x18, 0xb5, 0x25, 0x81, 0x2e, 0x0f,
+	}
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:989]:
```suggestion
	result := make([]byte, len(request))
	copy(result, request)
```
and now work on the `result`

File: pkg/network/usm/monitor_test.go [hunk 9]
```diff
@@ -1025,7 +1024,7 @@ func (s *USMHTTP2Suite) TestRawTraffic() {
 			// a single program.
 			numberOfSettingFrames: 100,
 			numberOfRequestFrames: 1,
-			requestBuffer:         request,
+			requestBuffer:         createHTTP2RequestByCase(caseDefault),
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1027]: Why? you shouldn't be touching the unrelated tests

File: pkg/network/usm/monitor_test.go [hunk 10]
```diff
@@ -964,30 +964,44 @@ var (
...
 	}
 
-	// http2 request to path /aaa
-	request = []byte{
-		0x00, 0x00, 0x37, 0x01, 0x04, 0x00, 0x00, 0x00, 0x03, 0x41, 0x8a, 0x08, 0x9d, 0x5c, 0x0b, 0x81,
-		0x70, 0xdc, 0x78, 0x0f, 0x0b, 0x83, 0x45, 0x83, 0x60, 0x63, 0x1f, 0x86, 0x5f, 0x8b, 0x1d, 0x75,
-		0xd0, 0x62, 0x0d, 0x26, 0x3d, 0x4c, 0x74, 0x41, 0xea, 0x5c, 0x01, 0x34, 0x50, 0x83, 0x9b, 0xd9,
-		0xab, 0x7a, 0x8d, 0xc4, 0x75, 0xa7, 0x4a, 0x6b, 0x58, 0x94, 0x18, 0xb5, 0x25, 0x81, 0x2e, 0x0f,
-	}
-
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:974]: restore

File: pkg/network/usm/monitor_test.go [hunk 11]
```diff
@@ -964,30 +964,44 @@ var (
...
 
-// createRawRequestFrame creates a raw request frame with the given streamID.
-func createRawRequestFrame(streamID int) []byte {
-	binary.BigEndian.PutUint16(request[7:9], uint16(streamID))
+type literalType int
+
+const (
+	caseWithoutIndexing literalType = iota
+	caseNeverIndexed
+	caseDefault
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:978]: redundant, remove

File: pkg/network/usm/monitor_test.go (hunk 8 above)
Comment (guyarb) [reply pkg/network/usm/monitor_test.go]: you're missing the copy

File: pkg/network/usm/monitor_test.go [hunk 12]
```diff
@@ -978,16 +978,31 @@ var (
...
+	result := make([]byte, len(request))
+	switch requestType {
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		result[22] = 0x05
+	// http2 request to path /aaa which never indexed (same as request but with different index changed from 0x45 to 0x15)
+	case caseNeverIndexed:
+		result[22] = 0x15
+	}
+	return result
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/usm/monitor_test.go:999]: https://github.com/DataDog/datadog-agent/pull/21913#discussion_r1444560086

**Review by guyarb — APPROVED** [review]

**Comment by guyarb** [comment]
please elaborate in the PR description and add the commit in grpc-tls PR that you took, and about the new 2 use cases we support now

File: pkg/network/protocols/http2/raw_traffic.go [hunk 13]
```diff
@@ -10,14 +10,17 @@ package http2
 import (
 	"bytes"
 	"fmt"
-
```
Comment (guyarb) [inline pkg/network/protocols/http2/raw_traffic.go:13]: revert

File: pkg/network/protocols/http2/raw_traffic.go [hunk 14]
```diff
@@ -10,14 +10,17 @@ package http2
...
 )
 
+const DynamicTableSize = 100
+
 // NewHeadersFrameMessage creates a new HTTP2 data frame message with the given header fields.
 func NewHeadersFrameMessage(headerFields []hpack.HeaderField) ([]byte, error) {
 	var buf bytes.Buffer
 	enc := hpack.NewEncoder(&buf)
+	// we set the max dynamic table size to 100 to be able to test different cases of literal header parsing.
+	enc.SetMaxDynamicTableSizeLimit(DynamicTableSize)
```
Comment (guyarb) [inline pkg/network/protocols/http2/raw_traffic.go:23]: I'm not sure this is right to do it everytime - maybe an optional parameter instead or split to functions

File: pkg/network/usm/monitor_test.go [hunk 15]
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
...
-	{Name: ":scheme", Value: "http"},
-	{Name: "content-type", Value: "application/json"},
-	{Name: "content-length", Value: "4"},
-	{Name: "accept-encoding", Value: "gzip"},
-	{Name: "user-agent", Value: "Go-http-client/2.0"},
+// generateTestHeaderFields generates a set of header fields that will be used for the tests.
+func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.HeaderField {
+	pathHeaderField := hpack.HeaderField{Name: ":path", Value: "/aaa", Sensitive: false}
+
+	// If the path is sensitive, we want to make sure that the path is longer than dynamicTableSize characters.
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:960]: The documentation does not make sense
if the path is sensitive, we want a longer path?

File: pkg/network/usm/monitor_test.go [hunk 16]
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
...
+	// Therefore, we want to make sure that the path is longer than 100 characters so that the path will not be indexed.
+	// ref: https://github.com/golang/net/blob/07e05fd6e95ab445ebe48840c81a027dbace3b8e/http2/hpack/encode.go#L140
+	if withoutIndexing {
+		pathHeaderField = hpack.HeaderField{Name: ":path", Value: "/" + strings.Repeat("a", usmhttp2.DynamicTableSize), Sensitive: true}
+	}
+
+	// If the path is sensitive, we are in a case where a path is never indexed
+	if pathNeverIndexed {
+		pathHeaderField = hpack.HeaderField{Name: ":path", Value: "/aaa", Sensitive: true}
+	}
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:971]: you're allowing passing both options as "true", which will not work for you

File: pkg/network/usm/monitor_test.go [hunk 17]
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
...
-	{Name: ":authority", Value: http2SrvAddr},
-	{Name: ":method", Value: "POST"},
-	{Name: ":path", Value: "/aaa"},
-	{Name: ":scheme", Value: "http"},
-	{Name: "content-type", Value: "application/json"},
-	{Name: "content-length", Value: "4"},
-	{Name: "accept-encoding", Value: "gzip"},
-	{Name: "user-agent", Value: "Go-http-client/2.0"},
+// generateTestHeaderFields generates a set of header fields that will be used for the tests.
+func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.HeaderField {
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:957]: You can add wrappers 

```suggestion
// generateTestHeaderFields generates a set of header fields that will be used for the tests.
func headersWithNeverIndexedPath() []hpack.HeaderField { return generateTestHeaderFields(...) }
func headersWithoutIndexingPath() []hpack.HeaderField { return generateTestHeaderFields(...) }
func defaultHeaders() []hpack.HeaderField { return generateTestHeaderFields(...) }
```

Comment (guyarb) [reply pkg/network/usm/monitor_test.go:957]: that way you dont need to pass `false, false` for most of the instances or `true, false` or `false, true`

**Review by guyarb — APPROVED** [review]

**Pull request merged**